js/woo.js

```diff
@@ -0,0 +1,1453 @@
+        const method = this.getSupportedMapping (marketType, {
+            'spot': 'v1PrivateGetClientTrades',
+        });
+        const response = await this[method] (this.extend (request, query));
+        // {
+        //     "success": true,
+        //     "meta": {
+        //         "records_per_page": 25,
+        //         "current_page": 1
+        //     },
+        //     "rows": [
+        //         {
+        //             "id": 5,
+        //             "symbol": "SPOT_BTC_USDT",
+        //             "order_id": 211,
+        //             "order_tag": "default",
+        //             "executed_price": 10892.84,
+        //             "executed_quantity": 0.002,
+        //             "is_maker": 0,
+        //             "side": "SELL",
+        //             "fee": 0,
+        //             "fee_asset": "USDT",
+        //             "executed_timestamp": "1566264290.250"
+        //         },
+        //         ...
+        //     ]
+        // }
+        const trades = this.safeValue (response, 'rows', []);
+        return this.parseTrades (trades, market, since, limit, params);
+    }
+
+    async fetchBalance (params = {}) {
+        await this.loadMarkets ();
+        const [ marketType, query ] = this.handleMarketTypeAndParams ('fetchBalance', undefined, params);
+        const method = this.getSupportedMapping (marketType, {
+            'spot': 'v2PrivateGetClientHolding',
+        });
+        const response = await this[method] (query);
+        //
+        // {
+        //     holding: [
+        //       {
+        //         token: 'USDT',
+        //         holding: '123.758485',
+        //         frozen: '22.0', // i.e. withdrawal
+        //         interest: '0.0',
+        //         outstanding_holding: '-56.4', // whatever in pending/limit orders
+        //         pending_exposure: '0.0', // this value is set when a pending order is waiting to get this token
+        //         opening_cost: '0.00000000',
+        //         holding_cost: '0.00000000',
+        //         realised_pnl: '0.00000000',
+        //         settled_pnl: '0.00000000',
+        //         fee_24_h: '0',
+        //         settled_pnl_24_h: '0',
+        //         updated_time: '1641370779'
+        //       },
+        //       ...
+        //     ],
+        //     success: true
+        // }
+        //
+        return this.parseBalance (response);
+    }
+
+    parseBalance (response) {
+        const result = {
+            'info': response,
+        };
+        const balances = this.safeValue (response, 'holding', []);
+        for (let i = 0; i < balances.length; i++) {
+            const balance = balances[i];
+            const code = this.safeCurrencyCode (this.safeString (balance, 'token'));
+            const account = this.account ();
+            account['total'] = this.safeString (balance, 'holding');
+            account['used'] = this.safeString (balance, 'outstanding_holding');
```

Comment:
I think `used` should be a `this.sum` in this case:

```JavaSciprt
const outstandingHolding = this.safeString (balance, 'outstanding_holding');
const frozen = this.safeString (balance, 'frozen');
const used = Precise.stringAdd (frozen, Precise.stringNeg (outstandingHolding));
```

In general, `total = used + free` should hold – let's ensure that.

Reply:
yes, that's done